core/src/main/scala/kafka/log/Log.scala

```diff
@@ -181,25 +191,38 @@ class Log(val dir: File,
         // if its a log file, load the corresponding log segment
         val start = filename.substring(0, filename.length - LogFileSuffix.length).toLong
         val indexFile = Log.indexFilename(dir, start)
+        val timeIndexFile = Log.timeIndexFilename(dir, start)
         val segment = new LogSegment(dir = dir,
                                      startOffset = start,
                                      indexIntervalBytes = config.indexInterval,
                                      maxIndexSize = config.maxIndexSize,
                                      rollJitterMs = config.randomSegmentJitter,
                                      time = time,
                                      fileAlreadyExists = true)
+        if (indexFile.exists() && timeIndexFile.exists()) {
 
-        if(indexFile.exists()) {
           try {
-              segment.index.sanityCheck()
+            segment.index.sanityCheck()
           } catch {
             case e: java.lang.IllegalArgumentException =>
-              warn("Found a corrupted index file, %s, deleting and rebuilding index...".format(indexFile.getAbsolutePath))
+              warn(s"Found a corrupted index file, deleting ${timeIndexFile.getAbsolutePath}, " +
+                s"${indexFile.getAbsolutePath} and rebuilding index...")
               indexFile.delete()
+              timeIndexFile.delete()
               segment.recover(config.maxMessageSize)
           }
-        }
-        else {
+
+          try {
+            segment.timeIndex.sanityCheck()
+          } catch {
+            case e: java.lang.IllegalArgumentException =>
+              warn(s"Found a corrupted time index file, deleting ${timeIndexFile.getAbsolutePath}, " +
+                  s"${indexFile.getAbsolutePath} and rebuilding index...")
+              indexFile.delete()
+              timeIndexFile.delete()
+              segment.recover(config.maxMessageSize)
+          }
+        } else {
```

Comment:
I looked at the commit history on this class a bit more. It seems that the existing logic is because of a bug introduced in KAFKA-2012. We should check the existence of the offset index file before instantiating a LogSegment, not after. Then, if the offset index is missing, we will rebuild the index. So, perhaps we can fix the logic here as the following. (1) If offset index is missing, rebuild both indexes. (2) If only the timeindex is missing, create an empty timeindex file (i.e., assuming that it's an old segment from pre-v10).